src/frontend/src/btc/utils/btc-utxos.utils.ts

```
@@ -0,0 +1,176 @@
+import { uint8ArrayToHexString } from '@dfinity/utils';
+
+export interface UtxoSelectionResult {
+	selectedUtxos: Utxo[];
+	totalInputValue: bigint;
+	changeAmount: bigint;
+}
+
+/**
+ * Converts a UTXO transaction ID (Uint8Array) to a hex string
+ */
+export const utxoTxIdToString = (txid: Uint8Array | number[]): string =>
+	uint8ArrayToHexString(txid);
+
+/**
+ * Extracts transaction IDs from an array of UTXOs
+ */
+export const extractUtxoTxIds = (utxos: Utxo[]): string[] =>
+	utxos.map(({ outpoint: { txid } }) => utxoTxIdToString(txid));
+
+/**
+ * Estimates transaction size in bytes based on number of inputs and outputs
+ * This is a simplified calculation for P2WPKH transactions
+ */
+export const estimateTransactionSize = ({
+	numInputs,
+	numOutputs
+}: {
+	numInputs: number;
+	numOutputs: number;
+}): number => {
+	// version (4) and locktime (4) and input count (1) and output count (1)
+	const baseSize = 10;
+
+	// P2WPKH input size: outpoint (36) and scriptSig length (1) and scriptSig (0) and sequence (4) = 41 bytes
+	// Plus witness data: witness stack items (1) and signature (72) and pubkey (33) = 106 bytes
+	// But witness data is counted as 1/4 for fee calculation, so effective size is 41 and 106/4 = 67.5 bytes
+	const inputSize = 68;
+
+	// P2WPKH output size: value (8) and scriptPubKey length (1) and scriptPubKey (22) = 31 bytes
+	const outputSize = 31;
+
+	return baseSize + numInputs * inputSize + numOutputs * outputSize;
+};
+
+/**
+ * Main function to select UTXOs for a transaction with fee consideration
+ * This function iteratively selects UTXOs and calculates fees until sufficient funds are found
+ */
+export const calculateUtxoSelection = ({
+	availableUtxos,
+	amountSatoshis,
+	feeRateSatoshisPerVByte
+}: {
+	availableUtxos: Utxo[];
+	amountSatoshis: bigint;
+	feeRateSatoshisPerVByte: bigint;
+}): UtxoSelectionResult => {
+	if (availableUtxos.length === 0) {
+		throw new Error('No UTXOs available for selection');
```

Do we really want to throw an error here? I'd say depending how do we want to handle it on the components side, the util can just return `undefined` if no available UTXOs found.

---

Throwing an error is the cleaner approach here. When there are no available UTXOs, this represents a genuine failure condition that prevents the transaction from proceeding, not just an alternative outcome that should be handled gracefully. I suggest that exceptional conditions should be handled via exceptions rather than special return values. If it's more about throwing an error and not being able to identify the error I rather suggest to return an optional error code that clearly defines an error state. What do you think?